app/controllers/users_controller.rb

```diff
@@ -0,0 +1,48 @@
+#
+class UsersController < ApplicationController
+  def new
+    @user = User.new
+  end
+
+  def index
+    @users = User.all
+  end
+
+  def show
+    @user = User.find(params[:id])
+  end
+
+  def create
+    @user = User.new(user_params)
+    if @user.save
+      session[:user_id] = @user.id
+      flash[:success] = 'Signed Up'
+      redirect_to @user
+    else
+      render 'static_pages/error'
+    end
+  end
+
+  def edit
+    @user = User.find(params[:id])
+    redirect_to root_path && return unless @user == current_user
```

Comment:
Theoretically I get why, but I am not sure as to the how. The 4.0 Guide doesn't spend much time with callbacks. I think I would need a before_validate callback, but am not totally confident. (Including so i don't forget)